utils/loading_cache.hh

```diff
@@ -290,6 +290,31 @@ public:
         });
     }
 
+    void put(const Key& k, value_type value) {
+        // We shouldn't be here if caching is disabled
+        assert(caching_enabled());
+
+        timestamped_val_ptr ts_val_ptr = _loading_values.get_or_insert(k, std::move(value));
+        // check again since it could have already been inserted and initialized
+        if (!ts_val_ptr->ready() && !ts_val_ptr.orphaned()) {
+            _logger.trace("{}: putting the value for the first time", k);
+
+            if (ts_val_ptr->size() > _max_size) {
+                throw entry_is_too_big();
+            }
+
+            ts_value_lru_entry* new_lru_entry = Alloc().template allocate_object<ts_value_lru_entry>();
+
+            // Remove the least recently used items if map is too big.
+            shrink();
```

Comment:
`shrink()` is technically not marked noexcept, so we risk leaking memory if it throws. But in practice I see that it never throws anyway, so I'll send a separate patch which sprays a few `noexcept`s in this code to be explicit about it.

Reply:
@psarna This probably because an Alloc().delete_object(...) is not `noexcept`

Reply:
@vladzcloudius It's not marked as such, but it's specified in the documentation that it doesn't throw: https://en.cppreference.com/w/cpp/memory/polymorphic_allocator/delete_object . It's a destruction path, so it makes sense for it to never throw

Reply:
True. However I wasn't checking the cppreference webpage text but rather a code of the corresponding class.

In C++10 I have it's still not marked as `noexcept`. 

A `noexcept` qualifier can make you more harm than good in such a situation (if something WOULD throw despite the text on the webpage you'll get your application `terminate`ed).
Any particular reason why you are so determined to mark `shrink` as `noexcept`?
@psarna

Reply:
It's either that or wrap parts of this code with try-catch, because we shouldn't leak memory. Since the code doesn't throw, I think that noexcept is more correct. I'm fine if you prefer to add try-catches, we just shouldn't leave the code as is.

Reply:
We can assume delete/destroy are all noexcept. It's impossible to write any program if they throw, since the clean up paths must call them.